base/MAPL_IO.F90

```diff
@@ -9291,4 +9331,148 @@ function get_fname_by_face(fname, face) result(name)
 
+   function check_flip(metadata,rc) result(flip)
+      type(FileMetadata), intent(inout) :: metadata
+      integer, optional, intent(out) :: rc
+      character(len=:), pointer :: positive
+      type(CoordinateVariable), pointer :: var
+      type (StringVariableMap), pointer :: vars
+      type (StringVariableMapIterator) :: var_iter
+      character(len=:), pointer :: var_name
+      logical :: isPresent
+      logical :: flip
+      type(Attribute), pointer :: attr => null()
+      class(*), pointer :: vpos
+
+      flip = .false.
+      vars => metadata%get_variables()
+      var_iter = vars%begin()
+      do while(var_iter /=vars%end())
+         var_name => var_iter%key()
+         var => metadata%get_coordinate_variable(trim(var_name))
+         if (associated(var)) then
+            if (index(var_name,'lev') .ne. 0 .or. index(var_name,'edge') .ne. 0) then
+               isPresent = var%is_attribute_present('positive')
+               if (isPresent) then
+                  attr => var%get_attribute('positive')
+                  _ASSERT(associated(attr),"restart file leve dim has no positive attribute")
+                  vpos => attr%get_value()
+                  select type(vpos)
+                  type is (character(*))
+                     positive => vpos
+                  class default
+                     _ASSERT(.false.,'units must be string')
+                  end select
+               else
+                  positive => null()
+               end if
+               if (positive == "up") then
+                  flip=.true.
+                  _RETURN(_SUCCESS)
+               end if
+            end if
+         end if
+         call var_iter%next()
+      enddo
+      _RETURN(_SUCCESS)
+   end function check_flip
+
+   subroutine flip_field(field,rc)
+      type(ESMF_Field), intent(inout) :: field
+      integer, intent(out), optional :: rc
+
+      integer :: status,rank
+      real(KIND=ESMF_KIND_R4), pointer :: ptr_r4(:,:,:)
+      real(KIND=ESMF_KIND_R8), pointer :: ptr_r8(:,:,:)
+      real(KIND=ESMF_KIND_R4), allocatable :: alloc_r4(:,:,:)
+      real(KIND=ESMF_KIND_R8), allocatable :: alloc_r8(:,:,:)
+      type(ESMF_TypeKind_Flag) :: tk
+      integer :: vloc,i,lb,ub,ii
+     
+      call ESMF_FieldGet(field,rank=rank,typeKind=tk,rc=status)
+      _VERIFY(status)
+      if (rank==3) then
+         call ESMF_AttributeGet(field,name="VLOCATION",value=vloc,rc=status)
+         _VERIFY(status)
+         if (vloc==MAPL_VLocationCenter .or. vloc==MAPL_VLocationEdge) then
+            if (tk == ESMF_TYPEKIND_R4) then
+               call ESMF_FieldGet(field,farrayPtr=ptr_r4,rc=status)
+               _VERIFY(status)
+               allocate(alloc_r4,source=ptr_r4)
+               lb = lbound(ptr_r4,dim=3)
+               ub = ubound(ptr_r4,dim=3)
+               ii=0
+               do i=lb,ub
+                  ptr_r4(:,:,i)=alloc_r4(:,:,ub-ii)
+                  ii=ii+1
+               enddo
```

Comment:
Someday we'll have templates and won't have to duplicate code for r4/r8 each time we turn around.  (Just venting.)